test/test_basic_logic.py

```diff
@@ -105,14 +105,142 @@ def test_sending_data(self):
         c.initiate_connection()
         c.send_headers(1, self.example_request_headers)
 
+        buffer = h2.frame_buffer.FrameBuffer(server=True)
+        buffer.max_frame_size = 65535
+        buffer.add_data(c.data_to_send())
+
         # Clear the data, then send some data.
         c.clear_outbound_data_buffer()
         events = c.send_data(1, b'some data')
         assert not events
+        data_to_send = c.data_to_send()
         assert (
-            c.data_to_send() == b'\x00\x00\t\x00\x00\x00\x00\x00\x01some data'
+            data_to_send == b'\x00\x00\t\x00\x00\x00\x00\x00\x01some data'
+        )
+
+        # sanity check on frames and flags.
+        buffer.add_data(data_to_send)
+        frames = list(buffer)
+        assert len(frames) == 3
+        settings_frame = frames[0]
+        header_frame = frames[1]
+        data_frame = frames[2]
+        assert isinstance(settings_frame, hyperframe.frame.SettingsFrame)
+        assert isinstance(header_frame, hyperframe.frame.HeadersFrame)
+        assert isinstance(data_frame, hyperframe.frame.DataFrame)
+
+        assert 'PADDED' not in data_frame.flags
+
+    def test_sending_data_with_padding(self):
+        """
+        Single data frames with padding are encoded correctly.
+        """
+        c = h2.connection.H2Connection()
+        c.initiate_connection()
+        c.send_headers(1, self.example_request_headers)
+
+        buffer = h2.frame_buffer.FrameBuffer(server=True)
+        buffer.max_frame_size = 65535
+        buffer.add_data(c.data_to_send())
+
+        # Clear the data, then send some data.
+        c.clear_outbound_data_buffer()
+        events = c.send_data(1, b'some data', pad_length=5)
+        assert not events
+        data_to_send = c.data_to_send()
+        assert data_to_send == (
+            b'\x00\x00\x0f\x00\x08\x00\x00\x00\x01'
+            b'\x05some data\x00\x00\x00\x00\x00'
         )
 
+        # sanity check on frames and flags
+        buffer.add_data(data_to_send)
+        frames = list(buffer)
+        assert len(frames) == 3
+        settings_frame = frames[0]
+        header_frame = frames[1]
+        data_frame = frames[2]
+        assert isinstance(settings_frame, hyperframe.frame.SettingsFrame)
+        assert isinstance(header_frame, hyperframe.frame.HeadersFrame)
+        assert isinstance(data_frame, hyperframe.frame.DataFrame)
+
+        assert 'PADDED' in data_frame.flags
+
```

Comment:
So there is a bunch of duplicate code around checking the output of the stream and whether there is a padding flag, and if so how much. It might be nice to factor that out to a helper function.